source/common/common/utility.h

```diff
@@ -288,4 +301,57 @@ class RegexUtil {
                                std::regex::flag_type flags = std::regex::optimize);
 };
 
+/**
+ * Maintains sets of numeric intervals. As new intervals are added, existing ones in the
+ * set are combined so that no overlapping intervals remain in the representation.
+ *
+ * Value can be any type that is comparable with <, ==, and >.
+ */
+template <typename Value> class IntervalSetImpl : public IntervalSet<Value> {
+public:
+  typedef typename IntervalSet<Value>::Interval Interval;
+
+  void insert(Value left, Value right) override {
+    if (left == right) {
+      return;
+    }
+    ASSERT(left < right);
+
+    auto left_pos = intervals_.lower_bound(Interval(left, left));
+    // upper_bound is exclusive, and we want to be inclusive.
+    auto right_pos = intervals_.upper_bound(Interval(right, right));
+    if (!intervals_.empty()) {
+      --right_pos;
+    }
+
+    if ((left_pos == intervals_.end()) || (right_pos == intervals_.end()) ||
+        (right < left_pos->first) || (right_pos->second < left)) {
+      // Fully disjoint. Simply insert.
+      intervals_.insert(Interval(left, right));
+    } else {
+      // Both bounds overlap.
+      left = std::min(left_pos->first, left);
+      right = std::max(right_pos->second, right);
+      ++right_pos; // erase is non-inclusive on upper bound.
+      intervals_.erase(left_pos, right_pos);
+      intervals_.insert(Interval(left, right));
+    }
+  }
+
+  std::vector<Interval> toVector() const override {
```

Comment:
It would be more readable to just write:

```
return std::vector<Interval>(intervals_.begin(), intervals_.end());
```

But, this might actually be less efficient because you don't get to `reserve`. Even if you do the construct and `reserve`, how about `out.insert(out.begin(), intervals_.begin(), intervals.end())` instead of manually writing the loop?

Reply:
https://stackoverflow.com/a/35359472 (pulling from the docs) indicates that since you're passing in a type of forward iterator, it should reserve for you, so you should be able to just one-line it.

Reply:
Nice (if it's easy, @jmarantz might want to try a benchmark too - I know there are some micro benchmarking tools he has used)

Reply:
I switched to using std::vector::insert rather than the loop, but kept the reserve(). I don't see how STL could know how much to reserve given two iterators without doing a test iteration.  However I will add a benchmark & ping again when I've got i.

Reply:
Added benchmark, which showed that (counterintuitively to me), the simpler
  return std::vector<Interval>(intervals_.begin(), intervals_.end());
turned out to be slightly faster for 3 intervals, and almost exactly the same speed for 50 disjoint intervals.

Lesson learned is -- actually -- just write the code in the simplest reasonable way, and then optimize if you can prove it helps.

Reply:
Nice! Good job `std::vector` implementation. I think the answer has something to do with the specifics of "Forward Iterators" - but my STL-fu isn't good enough to grok it.

Reply:
There’s a stdlib iteration distance function that they’re probably using http://en.cppreference.com/w/cpp/iterator/distance